### ISSUE TITLE
[Bug] Saving fails with NullReferenceException when a character is carrying inventory

### ISSUE THREAD

**Issue by koosemose** [TeamPorcupine/ProjectPorcupine#372]
It fails on Inventory.cs line 106; `writer.WriteAttributeString("X", tile.X.ToString());`. Obviously an inventory being carried by a character isn't going to have a tile. What would be the best way to go about this, just skip saving an inventory without a tile and assume the character carrying it (or whatever other nontile location it may be capable of being in in the future) will save it?

**Comment by Mizipzor** [comment]
You already know how to see which line it was but including the rest of the stack trace can be super useful. :)

**Comment by koosemose** [comment]
Well I already know what the problem is, I'm just looking for advice on the best logic to use in the fix. And the rest of the stacktrace isn't really that useful since it's only going to be called from one place, and I know why exactly it was called, which is much more clear than trying to parse the stacktrace... but if you really want it.

`NullReferenceException: Object reference not set to an instance of an object
Inventory.WriteXml (System.Xml.XmlWriter writer) (at Assets/Scripts/Models/Inventory.cs:107)
World.WriteXml (System.Xml.XmlWriter writer) (at Assets/Scripts/Models/World.cs:455)
System.Xml.Serialization.XmlSerializationWriter.WriteSerializable (IXmlSerializable serializable, System.String name, System.String ns, Boolean isNullable, Boolean wrapped)
System.Xml.Serialization.XmlSerializationWriter.WriteSerializable (IXmlSerializable serializable, System.String name, System.String ns, Boolean isNullable)
System.Xml.Serialization.XmlSerializationWriterInterpreter.WriteObject (System.Xml.Serialization.XmlTypeMapping typeMap, System.Object ob, System.String element, System.String namesp, Boolean isNullable, Boolean needType, Boolean writeWrappingElem)
System.Xml.Serialization.XmlSerializationWriterInterpreter.WriteRoot (System.Object ob)
System.Xml.Serialization.XmlSerializer.Serialize (System.Object o, System.Xml.Serialization.XmlSerializationWriter writer)
System.Xml.Serialization.XmlSerializer.Serialize (System.Xml.XmlWriter writer, System.Object o, System.Xml.Serialization.XmlSerializerNamespaces namespaces)
UnityEngine.EventSystems.EventSystem:Update()
`

**Comment by Mizipzor** [comment]
It wasnt for your sake rather for everyone else that finds this issue and wants to understand whats going on. :)

**Comment by koosemose** [comment]
Ahh... though I really feel like in this case the stacktrace isn't useful... except maybe as practice for someone to see if they can arrive at the problem with just the stacktrace, as though a minimal bugreport had been filed (Though a proper "minimal bugreport" would be "IT"S BREOKKE FIXXX IT!!!"... for which the proper solution is something blunt...)

**Comment by Mizipzor** [comment]
I agree that the stack trace wasnt that useful here since the call is not coming from our code (but I didnt know that initially). 

There are however deeper problems here. If an object that is carried doesnt have a tile, how would it be picked up again when the game loads? It doesnt save who it is carried by. Or is it the carrier that saves that? Maybe it should have a position that corresponds to the carriers position.

I also think there are some architectural concerns here. Is should probably create a separate issue for this. But in short, there is a mix between the **definition** of an item and an **instance** of that item. It should be two classes. The definition holds data that is shared across all instances, like `MaxStackSize` and later stuff like weight or whatever we add. The instances then holds all the information that cannot be shared, like position. Once we have lots of information defining an item it doesnt make sense to copy that info across all definitions.

**Comment by koosemose** [comment]
Currently it appears that the character doesn't save his carried inventory, so carried inventory would be lost, which certainly makes for a more complex fix. I think, for now at the very least, it makes sense for the inventory to be saved with the character. However, the problem is that either character will need it's own version of the inventory WriteXml (and probably ReadXml), which seems just wrong, or inventory will need separate versions of them which primarily would be used by Character to write the inventory it's holding. It can't just do a simple skip based on tile being null, because it needs to not be written when the inventory writer comes through, but would only need to skip trying to write x and y when the character writer comes through. This has suddenly become much more complex than I feel capable of sorting through at the moment.

**Comment by Mizipzor** [comment]
Thats fine, just because you reported the issue doesnt mean you would have to fix it yourself. :) Youre absolutely right that the proper fix for this will be more involved. 

If you still want to help how about coming up with a good way to reproduce the problem? I dont know of the best way to spawn a worker and an item, have the worker pick up that item and stand still with it, save and then load and then verify. Seems like it would be a lot of steps if it would happen naturally in the game.

**Comment by koosemose** [comment]
A Mining drone and at least 2 smelters works fairly well for now (more smelters will make it easier to catch), due to the current bug with the smelter where they will take from the input stack of one smelter to feed another. I tended to have 5 smelters running when I would come across it naturally.

Alternatively load a game (so it generates steel plates) and set up a stockpile a decent distance from the plates, the farther it is the easier it will be to catch. 

And I tend to like to fix the bugs I find myself, mostly out of impatience, that way (from my, possibly warped, perception) I just need to wait on the PR to be accepted, but for someone else I also need to wait for them to actually do it.

**Comment by bjubes** [comment]
I think having the character save his own inventory is the best fix since it follows the hierarchy of 
world > tiles > inventory
character > inventory
as opposed to the inventory saving whos holding it, thus reversing that flow.

**Comment by kd7uiy** [comment]
Is this still an issue?

**Comment by vogonistic** [comment]
I'm closing this. Reopen if it's still a bug.
